program-runtime/src/mem_pool.rs

```diff
@@ -0,0 +1,146 @@
+use {
+    solana_compute_budget::{
+        compute_budget::{MAX_CALL_DEPTH, MAX_INSTRUCTION_STACK_DEPTH, STACK_FRAME_SIZE},
+        compute_budget_processor::{MAX_HEAP_FRAME_BYTES, MIN_HEAP_FRAME_BYTES},
+    },
+    solana_rbpf::{aligned_memory::AlignedMemory, ebpf::HOST_ALIGN},
+    std::array,
+};
+
+trait Reset {
+    fn reset(&mut self);
+}
+
+struct Pool<T: Reset, const SIZE: usize> {
+    items: [Option<T>; SIZE],
+    next_empty: usize,
+}
+
+impl<T: Reset, const SIZE: usize> Pool<T, SIZE> {
+    fn new(items: [T; SIZE]) -> Self {
+        Self {
+            items: items.map(|i| Some(i)),
+            next_empty: SIZE,
+        }
+    }
+
+    fn len(&self) -> usize {
+        SIZE
+    }
+
+    fn get(&mut self) -> Option<T> {
+        if self.next_empty == 0 {
+            return None;
+        }
+        self.next_empty = self.next_empty.saturating_sub(1);
+        self.items
+            .get_mut(self.next_empty)
+            .and_then(|item| item.take())
+    }
+
+    fn put(&mut self, mut value: T) -> bool {
+        self.items
+            .get_mut(self.next_empty)
+            .map(|item| {
+                value.reset();
```

Comment:
Do we have any tests to check that memory is clean when the guest enters?

Reply:
Several existing tests fail if the memory isn't reset:

```
failures:
    assert_instruction_count
    test_cpi_deprecated_loader_realloc
    test_program_sbf_invoke_sanity
    test_program_sbf_realloc
    test_program_sbf_realloc_invoke
    test_program_sbf_sanitytest
result: FAILED. 32 passed; 6 failed; 1 ignored; 0 measured; 0 filtered out; finished in 3.63s
```